TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs

```diff
@@ -936,16 +937,52 @@ public static void CheckStatement(Node node)
 
     public class ProgramChecker
     {
-        public static void OnNode(Program node)
+        public static void OnNode(Program node, TypeCobolOptions typeCobolOptions)
         {
-            node.SetFlag(Node.Flag.MissingEndProgram, !(node.Children.LastOrDefault() is End));
-
-            if (node.IsFlagSet(Node.Flag.MissingEndProgram))
+            //var lastChild = node.Children.LastOrDefault();
+            //if (lastChild is End end)
+            //{
+            //    node.SetFlag(Node.Flag.MissingEndProgram, false);
+            //}
+            //else
+            //{
+            //    node.SetFlag(Node.Flag.MissingEndProgram, true);
+            //    DiagnosticUtils.AddError(node, "\"END PROGRAM\" is missing.", typeCobolOptions.CheckEndProgram.GetMessageCode());
+            //}
+
+            var lastChild = node.Children.LastOrDefault();
+            if (lastChild is End end)
             {
-                DiagnosticUtils.AddError(node,
-                    "\"END PROGRAM\" is missing.", MessageCode.Warning);
+                node.SetFlag(Node.Flag.MissingEndProgram, false);
+                if (typeCobolOptions.CheckEndProgram.IsActive)
+                {
+                    var programEnd = (ProgramEnd)end.CodeElement;
+                    if (programEnd.ProgramName?.Name == null)
+                    {
+                        DiagnosticUtils.AddError(end, "\"PROGRAM END\" should have a program name.", typeCobolOptions.CheckEndProgram.GetMessageCode());
+                    }
+                    else
+                    {
+                        if (!node.Name.Equals(programEnd.ProgramName.Name, StringComparison.OrdinalIgnoreCase))
+                        {
+                            DiagnosticUtils.AddError(end, "Program name not matching \"PROGRAM END\".", typeCobolOptions.CheckEndProgram.GetMessageCode());
+                        }
+                    }
+                }
+            }
+            else
+            {
+                node.SetFlag(Node.Flag.MissingEndProgram, true);
+                if (typeCobolOptions.CheckEndProgram.IsActive)
+                {
+                    if (node.IsMainProgram && node.GetChildren<Program>().Count == 0)
```

Comment:
Il y a la propriété `NestedPrograms` qui te donne cela plus facilement. Il ne faut pas aussi vérifier que c'est le seul programme du fichier (i.e. pas de stacked) ?

Reply:
Updated, but a question remains:

- `IsMainProgram` is only true for the first program in the source file.
- `IsStacked` is only true for programs that are children of root, but **not** the main program.

Is that intended?
Given the fact that there is 2 different booleans, isn't the main program supposed to have `IsStacked` to true when there is more than 1 program child of root?

Reply:
Yes it is intended. here `Stacked` means 'root program of source file but not the first'...

Reply:
If that is the intended behaviour then OK.